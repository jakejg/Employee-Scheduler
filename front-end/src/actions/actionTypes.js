export const LOAD_JOBS = "LOAD_JOBS";
export const ADD_JOBS = "ADD_JOB";