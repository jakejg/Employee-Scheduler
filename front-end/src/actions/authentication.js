import axios from 'axios';
import { 
    ADD_OR_REMOVE_TOKEN
 } from './actionTypes';
import {addCompanyToAPI} from '../helpers/JobApi'
import {BASE_URL} from '../config';


export const register = (data) => {
    
    return async (dispatch) => {
        try {
            // designate admin user
            data.is_admin = true
            // create admin user
       
            const res = await axios.post(`${BASE_URL}/auth/register`, data);
       
            dispatch(addOrRemoveToken(res.data.token));
        }
        catch(e) {
            console.log(e)
        }
    }
}

export const login = (data) => {
    return async (dispatch) => {
        try {
            let res = await axios.post(`${BASE_URL}/auth/login`, data);
            dispatch(addOrRemoveToken(res.data.token));
        }
        catch(e) {
            // console.log(e.response.data.message)
            return e.response.data.message
        }
    }
}

export const addOrRemoveToken = (token) => {
    return {type: ADD_OR_REMOVE_TOKEN, token}
}
