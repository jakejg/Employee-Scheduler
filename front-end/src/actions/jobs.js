import { 
    LOAD_JOBS,
    EDIT_JOB, 
    ADD_JOB,
    EDIT_JOB_STAFF,
    DELETE_JOB
 } from './actionTypes';
import {JobAPI} from '../helpers/JobApi';

export const loadJobsFromAPI = (comp_id) => {
    return async (dispatch) => {
        try {
            const jobData = await JobAPI.loadJobs(comp_id)
            dispatch(loadJobs(jobData));
        }
        catch(e) {
            console.log(e)
        }
    }
}

export const loadJobs = (jobs) => {
    return {type: LOAD_JOBS, jobs}
}

export const addJobOnAPI = (jobToAdd) => {

    return async (dispatch) => {
        try {
            let job = await JobAPI.addJob(jobToAdd)
            dispatch(addJob(job.id, job))
            return {message: "Job created Successfully", severity: "success"}
        }
        catch(e) {
            console.log(e)
        }
    }
}

export const addJob = (id, job) => {
    return {type: ADD_JOB, id, job}
}


export const getJobFromAPI = (ID) => {
    return async (dispatch) => {
        try {
            let job = await JobAPI.getJob(ID)
            dispatch(editJob(job.id, job))
        }
        catch(e) {
            console.log(e)
            return e.message
        }
    }
}

export const editJobOnAPI = (id, changedProps) => {
    return async (dispatch) => {
        try {
            let job = await JobAPI.editJob(id, changedProps)
            console.log(job)
            dispatch(editJob(job.id, job))
        }
        catch(e) {
            console.log(e)
            return e.message
        }
    }
}

export const editJob = (id, job) => {
    return {type: EDIT_JOB, id, job}
}

export const addStaffToJobOnAPI = (jobId, staffId) => {
    return async (dispatch) => {
        try {
            let data = await JobAPI.addStaffToJob(jobId, staffId)
            console.log(data)
            dispatch(editJobStaff(jobId, data.staffList, data.staff_filled))
            
        }
        catch(e) {
            console.log(e)
        }
    }
}

export const removeStaffFromJobOnAPI = (jobId, staffId) => {
    return async (dispatch) => {
        try {
            let data = await JobAPI.removeStaffFromJob(jobId, staffId)
            dispatch(editJobStaff(jobId, data.staffList, data.staff_filled))
            
        }
        catch(e) {
            console.log(e)
        }
    }
}

export const editJobStaff = (id, staff, staff_filled) => {
    return {type: EDIT_JOB_STAFF, id, staff, staff_filled}
}

export const deleteJobOnAPI = (id) => {
    return async (dispatch) => {
        try {
            let msg = await JobAPI.deleteJob(id)
            dispatch(deleteJob(id))
        }
        catch(e) {
            console.log(e)
            return e.message
        }
    }
}

export const deleteJob = (id) => {
    return {type: DELETE_JOB, id}
}