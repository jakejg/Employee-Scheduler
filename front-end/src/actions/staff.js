import { 
    LOAD_STAFF,
    EDIT_STAFF,
    ADD_STAFF
 } from './actionTypes';
import { StaffAPI } from '../helpers/StaffApi';

export const loadStaffFromAPI = (comp_id) => {
    return async (dispatch) => {
        try {
            let staffData = await StaffAPI.loadStaff(comp_id);
            dispatch(loadStaff(staffData));
        }
        catch(e) {
            console.log(e)
        }
    }
}

export const loadStaff = (staff) => {
    return {type: LOAD_STAFF, staff}
}


export const getStaffFromAPI = (ID) => {
    return async (dispatch) => {
        try {
            let staff = await StaffAPI.getStaff(ID)
        
            dispatch(editStaff(staff.id, staff))
            }
        catch(e) {
            console.log(e)
        }
    }
}

export const editStaff = (id, staff) => {
    return {type: EDIT_STAFF, id, staff}
}

export const addStaffOnAPI = (staffToAdd) => {
   
    return async (dispatch) => {
        try {
            const staff = await StaffAPI.addStaff(staffToAdd) ;
            dispatch(addStaff(staff.id, staff))
        }
        catch(e) {
            console.log(e)
        }
    }
}

export const addStaff = (id, staff) => {
    return {type: ADD_STAFF, id, staff}
}