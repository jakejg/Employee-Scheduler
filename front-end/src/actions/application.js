import {CHANGE_DRAWER, CHANGE_NO_JOBS, CHANGE_NO_STAFF} from './actionTypes';

export const changeDrawer = () => {
    return {type: CHANGE_DRAWER}
}

export const changeNoJobs = () => {
    return {type: CHANGE_NO_JOBS}
}

export const changeNoStaff = () => {
    return {type: CHANGE_NO_STAFF}
}