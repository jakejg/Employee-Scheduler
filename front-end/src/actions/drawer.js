import {CHANGE_DRAWER} from './actionTypes';

export const changeDrawer = () => {
    return {type: CHANGE_DRAWER}
}