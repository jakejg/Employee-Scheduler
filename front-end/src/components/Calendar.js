import React, {useEffect} from 'react';
import Timeline from 'react-calendar-timeline'
import './styles/Calendar.css';
import { useSelector, useDispatch } from 'react-redux';
import {Paper, Box, Container, Typography, makeStyles, List, ListItem, ListItemText } from '@material-ui/core';
import moment from 'moment'
import {useHistory} from 'react-router-dom';
import {editJob} from '../actions/jobs';
import { loadJobsFromAPI, editJobOnAPI } from '../actions/jobs';
import { decode } from 'jsonwebtoken';

const useStyles = makeStyles(() => ({
    instructions: {
        margin: '5px'
    },
    calendarContainer: {
        borderRadius: '10px'
    }
   
}))

const Calendar = () => {
    const jobs = useSelector(state => state.jobs);
    const token = useSelector(state => state.application.token);
    const classes = useStyles();
    const history = useHistory();
    const dispatch = useDispatch();

    useEffect(() => {
        const getData = async () => {
            const { comp_id } = decode(token);
            dispatch(loadJobsFromAPI(comp_id));
        }
        getData();
    }, [dispatch])

    // calender takes two props an array of events/item objects and array of grouping
    let items = Object.keys(jobs).map(id => jobs[id]);
    let groups = Object.keys(jobs).map(id => ({id: +id , title:jobs[id].title}));
    
    //groups can't be empty on the first render or it will cause dispaly issues
    if (!groups.length) groups = [{}];

    const handleDoubleClick = (itemId, e, time) => {
        history.push(`/job/${itemId}`)
    }
    const handleMove = (itemId, newStartTime, newGroupOrder) =>{
        let job = jobs[itemId];
       
        let {start_time, end_time } = job;
        job.end_time = moment(newStartTime + (end_time - start_time));
        job.start_time = moment(newStartTime);
        
        // change in database
        dispatch(editJobOnAPI(itemId, job))
        //change redux state so calendar re-renders smoothly
        dispatch(editJob(itemId, job))
    }
    

return (
    <Container>
        <Paper className={classes.instructions}>
            <List>
                <ListItem>
                    <ListItemText>
                        Click on a job to select it
                    </ListItemText>
                </ListItem>
                <ListItem>
                    <ListItemText>
                        Double click on a job to view details
                    </ListItemText>
                </ListItem>
            </List>
            <Typography></Typography>
        </Paper>
        <Paper elevation={4} className={classes.calendarContainer}>
        <Timeline
          groups={groups}
          items={items}
          canMove={true}
          itemHeightRatio={0.75}
          lineHeight={40}
          onItemDoubleClick={handleDoubleClick}
          onItemMove={handleMove}
          defaultTimeStart={moment().add(-30, 'day')}
          defaultTimeEnd={moment().add(30, 'day')}
        />
        </Paper>
    </Container>
)
}
export default Calendar;

