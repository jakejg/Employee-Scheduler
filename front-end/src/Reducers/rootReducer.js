import jobs from './jobsReducer';
import staff from './staffReducer';
import { combineReducers } from "redux";

export default combineReducers({jobs, staff})