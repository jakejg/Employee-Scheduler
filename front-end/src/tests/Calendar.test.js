import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import {Provider} from 'react-redux'
import Calendar from '../components/Calendar';
import {createStore, applyMiddleware} from "redux";
import rootReducer from "../reducers/rootReducer";
import { MemoryRouter } from "react-router";
import { loadJobs } from '../actions/jobs';
import { addOrRemoveToken } from '../actions/authentication';
import thunk from 'redux-thunk'
import moment from 'moment';

const jobs = {
    1 : {
        start_time : moment().add(-10, 'day'),
        end_time : moment().add(10, 'day'),
        group : 1,
        staff: [],
        staff_needed: 1,
        notes: "Test trip",
        status: 'under',
        title : '20 Day River'
    },
    2 : {
        start_time : moment().add(-5, 'day'),
        end_time : moment().add(-10, 'day'),
        group : 2,
        staff: [],
        status: 'under',
        title : '15 Day Mountain'
    }
}

const store = createStore( rootReducer, applyMiddleware(thunk));
store.dispatch(loadJobs(jobs)); 
store.dispatch(addOrRemoveToken("eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6IkRlbW8iLCJpc19hZG1pbiI6dHJ1ZSwiY29tcF9pZCI6MSwiaWF0IjoxNTk3NzY5NTYyfQ.Wm2z3GCj-m6DDWElCIBCuRxqBbc3_0QhJzeorsbz6E0")) ;

it('renders without crashing', () => {
   
  render(
    <MemoryRouter>
      <Provider store={store}>
      <Calendar />
      </Provider>
    </MemoryRouter>
)
});

it('matches snapshot', () => {
    const { asFragment } = render(
        <MemoryRouter>
            <Provider store={store}>
                <Calendar />
            </Provider>
        </MemoryRouter>
    )
    expect(asFragment()).toMatchSnapshot();
})


it('renders jobs on the calendar', () => {
    const { getAllByText } = render(
        <MemoryRouter>
            <Provider store={store}>
                <Calendar />
            </Provider>
        </MemoryRouter>
    )
    expect(getAllByText('20 Day River')[1]).toBeInTheDocument();
    expect(getAllByText('15 Day Mountain')[1]).toBeInTheDocument();
})

