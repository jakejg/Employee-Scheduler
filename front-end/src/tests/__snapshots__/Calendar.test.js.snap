// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg"
  >
    <div
      class="MuiPaper-root makeStyles-instructions-3 MuiPaper-elevation1 MuiPaper-rounded"
    >
      <ul
        class="MuiList-root MuiList-padding"
      >
        <li
          class="MuiListItem-root MuiListItem-gutters"
        >
          <div
            class="MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
            >
              Click on a job to select it
            </span>
          </div>
        </li>
        <li
          class="MuiListItem-root MuiListItem-gutters"
        >
          <div
            class="MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
            >
              Double click on a job to view details
            </span>
          </div>
        </li>
      </ul>
      <p
        class="MuiTypography-root MuiTypography-body1"
      />
    </div>
    <div
      class="MuiPaper-root makeStyles-calendarContainer-4 MuiPaper-elevation4 MuiPaper-rounded"
    >
      <div
        class="react-calendar-timeline "
      >
        <div
          class="rct-header-root"
          style="display: flex; width: 100%;"
        >
          <div
            style="width: 150px;"
          />
          <div
            class="rct-calendar-header"
            style="overflow: hidden;"
          >
            <div
              style="height: 30px; position: relative; width: -450px;"
            >
              <div
                class="rct-dateHeader rct-dateHeader-primary"
                style="left: 351.29734285300924px; width: -915px; position: absolute;"
              >
                <span>
                  20
                </span>
              </div>
            </div>
            <div
              style="height: 30px; position: relative; width: -450px;"
            >
              <div
                class="rct-dateHeader "
                style="left: 351.29734285300924px; width: -915px; position: absolute;"
              >
                <span>
                  20
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          class="rct-outer"
          style="height: 80px;"
        >
          <div
            class="rct-sidebar"
            style="width: 150px; height: 80px;"
          >
            <div
              style="width: 150px;"
            >
              <div
                class="rct-sidebar-row rct-sidebar-row-even"
                style="height: 40px; line-height: 40px;"
              >
                20 Day River
              </div>
              <div
                class="rct-sidebar-row rct-sidebar-row-odd"
                style="height: 40px; line-height: 40px;"
              >
                15 Day Mountain
              </div>
            </div>
          </div>
          <div
            class="rct-scroll"
            style="width: -150px; height: 100px; cursor: default; position: relative;"
          >
            <div
              style="position: absolute; left: 0px; right: 0px; top: 0px; bottom: 0px;"
            >
              <div
                class="rct-vertical-lines"
              >
                <div
                  class="rct-vl"
                  style="pointer-events: none; top: 0px; left: 351.29734285300924px; width: -915px; height: 80px;"
                />
              </div>
              <div
                class="rct-horizontal-lines"
              >
                <div
                  class="rct-hl-even "
                  style="width: -450px; height: 40px;"
                />
                <div
                  class="rct-hl-odd "
                  style="width: -450px; height: 40px;"
                />
              </div>
              <div
                class="rct-items"
              >
                <div
                  class="rct-item "
                  style="font-size: 12px; color: white; cursor: pointer; background: rgb(33, 150, 243); border: 1px solid #1a6fb3; z-index: 80; position: absolute; box-sizing: border-box; left: -212.49997517361112px; top: 45px; width: 12.5px; height: 30px; line-height: 30px;"
                  title="20 Day River"
                >
                  <div
                    class="rct-item-content"
                    style="max-height: 30;"
                  >
                    20 Day River
                  </div>
                </div>
                <div
                  class="rct-item "
                  style="font-size: 12px; color: white; cursor: pointer; background: rgb(33, 150, 243); border: 1px solid #1a6fb3; z-index: 80; position: absolute; box-sizing: border-box; left: -212.49997517361112px; top: 45px; width: 12.5px; height: 30px; line-height: 30px;"
                  title="15 Day Mountain"
                >
                  <div
                    class="rct-item-content"
                    style="max-height: 30;"
                  >
                    15 Day Mountain
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
